network.js

```diff
@@ -48,45 +48,71 @@ module.exports = {
   callFetch(dest, opts) {
     return fetch(dest, opts);
   },
-  downloadFile(url) {
-    downloadStats[url] = {tries: 0, bytesExpected: 0, bytesReceived: 0};
+  downloadFile(originalUrl) {
+    var resolve, reject,
+    savePath = path.join(platform.getTempDir(), url.parse(originalUrl).pathname.split("/").pop());
 
-    function tryDownload(resolve, reject) {
-      var tempPath = path.join(platform.getTempDir(), urlParse(url).pathname.split("/").pop()),
-      file = fs.createWriteStream(tempPath);
+    downloadStats[originalUrl] = {tries: 0, bytesExpected: 0, bytesReceived: 0};
 
-      downloadStats[url].tries += 1;
+    return new Promise((res, rej)=>{
+      resolve = res;
+      reject = rej;
+
+      setTimeout(()=>{
+        reject({ message: "Request timed out", error: originalUrl });
+      }, 1000 * 60 * 20);
+
+      tryDownload(originalUrl);
+    });
+
+    function tryDownload(downloadUrl) {
+      var file = fs.createWriteStream(savePath);
+
+      downloadStats[originalUrl].tries += 1;
 
       file.on("error", (err)=>{
         reject({ message: "Error creating temporary download file", error: err });
       });
 
-      log.debug("Downloading " + url + " try " + downloadStats[url].tries);
+      log.debug("Downloading " + originalUrl + " try " + downloadStats[originalUrl].tries);
 
-      var req = http.get(url, (res)=>{
-        if(res.statusCode === 404) {
-          reject({ message: "File not found", error: res.statusCode });
-        }
-        else if(res.statusCode < 200 || res.statusCode >= 300) {
+      var req = http.get(downloadUrl, (res)=>{
+        if (isRedirect(res.statusCode)) {
+          if (downloadStats[originalUrl].tries === maxRetries) {
+            reject({message: "Too many download attempts"});
+            return;
+          }
+          
+          if (!res.headers.location) {
+            reject({message: "Missing location header at " + originalUrl});
+            return;
+          }
+
+          tryDownload(url.resolve(downloadUrl, res.headers.location));
+          return;
+        } else if(res.statusCode < 200 || res.statusCode >= 300) {
           reject({ message: "Error downloading file", error: res.statusCode });
+          return;
```

Comment:
We could.  That's how I had it initially but I thought it might be more clear this way because we don't actually care about the return value of the reject function.  We're not chaining the promise.